qmail-qfilter.8
.\" vim: tw=75
.TH qmail-qfilter 8
.SH NAME
qmail-qfilter \- front end for qmail-queue that does filtering
.SH SYNOPSIS
.B qmail-qfilter
.I filter
[
.I -- filter ...
]
.SH DESCRIPTION
\fBqmail-qfilter\fR acts as a \fBqmail-queue\fR frontend to execute message
filters. It is invoked by setting the \fBQMAILQUEUE\fR environment variable
\fBqmail-qfilter\fR sends the message text through each of the filter
commands named on the command line and in the \fIqfilters\fR control
file (allowing finer control, see below). Each filter is run separately, with
standard input opened to the input email, and standard output opened to a
new temporary file that will become the input to either the next filter, or
\fBqmail-queue\fR. If the filter does not modify the message it passes
unchanged to the next step.  It also makes the envelope available to each
filter as file descriptor 3. File descriptor 4 is opened to a new temporary
file for the modified envelope, allowing the filter to modify the envelope
or the message. If the filter does not modify the envelope, the envelope
remains unchanged for either the next filter or \fBqmail-queue\fR. This
provides compatibility for existing filters that do not know about the
envelope. \fBqmail-qfilter\fR also opens up file descriptor 5 to a
temporary file. If this file is empty after all the filters have executed,
its contents are read and used to specify a program to execute in place of
\fBqmail-queue\fR. Each filter on the command line is separated by
\fB--\fR.

.SH "RETURN VALUES"
Returns 51 (out of memory), 53 (write error), or 81 (internal error)
if it can't create the temporary files or has problems executing the
filters.

Returns 79 (bad envelope data) if it can't read or parse the envelope
data.

\fBqmail-qfilter\fR returns exit code of filter, if the filter returns
anything other than 0 or 99. If your filter exits 88, you can put your
custom error string on standard error. This enables client like
\fBqmail-smtpd\fR to report the error. The custom error string can be
prefixed with \fBD\fR for permanent error and \fBZ\fR to indicate temporary
error. If a filter returns 99, \fBqmail-qfilter\fR returns 0 immediately
without running any other filters. If your filter returns 31, the mail will
be permanently rejected. If your filter returns 71, the mail will be
temporarily rejected.

.SH CONTROL FILES

.TP 5
.I qfilters
This control file can be used to setup additional filters. Lines in this
file have the following format:

.EX
\fBdomain:filter\fR
  or
\fBdomain:filter:envstr\fR
.EE

\fIdomain\fR refers to the envelope sender domain. It can be an exact match,
wildcard or a regular expression (if QREGEX environment variable is defined).
If there are multiple lines that match \fIdomain\fR, all matching filters
will be applied.

\fIfilter\fR is the filter command to run (passed through /bin/sh).

\fIenvstr\fR mentioned above is a string of comma-separated environment
variable additions or removals.

Example entry:

.EX
*:/var/qmail/bin/surbfilter:CACHELIFTIME=,SURBL_DEBUG_FN=/dev/null
.EE

.SH ENVIRONMENT
For compatibility with previous versions, \fBqmail-qfilter\fR sets
\fBQMAILUSER\fR and \fBQMAILHOST\fR to the user and host portions of the
envelope sender address, and unsets \fBQMAILNAME\fR. It also sets
\fBQMAILRCPTS\fR to the list of envelope recipients, each followed by a
newline.
.P
It also sets \fBENVSIZE\fR to the size of the envelope, \fBMSGSIZE\fR to
the length of the message, and \fBNUMRCPTS\fR to the number of recipients.
These values are updated before each filter is run.
.P
If \fBQQF_QMAILQUEUE\fR is set, that value is used in
place of \fBqmail-queue\fR.

.SH "SEE ALSO"
qmail-queue(8),
spawn-filter(8)

.SH NOTES
\fB$QMAILQUEUE\fR is deliberately not used to override \fBqmail-queue\fR in
order to avoid recursive loops with configurations that set
\fB$QMAILQUEUE\fR.

.SH WARNINGS
If you are using \fBqmail-inject\fR -n as one of the filters, you may want
to unset \fBMAILUSER\fR, \fBUSER\fR, and \fBLOGNAME\fR by using

.EX
.B env -u QMAILNAME -u MAILNAME -u NAME qmail-inject -n
.EE

as the command to invoke qmail-inject. Note that some the \fBenv\fR command
with some OS's doesn't support the \fB-u\fR option.
.P
A message with an excessive number of recipients (more than 64K bytes of
recipient data on Linux) will cause execution of the filter programs to
fail, and for the message to be rejected.
.P
The same temporary file is reused for file descriptor 5 for each filter.
Make sure each filter writes a trailing ASCII NUL byte following the
program name, as multiple filters could otherwise overwrite the value in
undesireable ways.


CHANGELOG
Changelog

20240926 Roberto Puzzanghera
         - code poliched from several errors/warnings against latest GCC 14.2
         - merged auth-tls-forcetls patch v. 2024.07.10

-------------------------------------------------------------------------------------

20071130 version: netqmail 1.06
20071130 legal: qmail-1.03 is now in the public domain
20051103 doc: dot-qmail.9 updated for changed (19980613) conf-patrn default, tnx ADM
20040121 version: netqmail 1.05
20040121 code: qmail-smtpd is protected from exceedingly long (eg 2GB)
         header lines
20040121 code: qmail_lspawn, qmail-newmrh, qmail-newu, and qmail-rspawn
         are protected from misbehaving on hosts where the size of an
         integer is not the same as the size of a character pointer
         (eg 64 bit hosts with 32 bit ints)
20031027 version: netqmail 1.04.
20031027 doc: INSTALL points to http://lifewithqmail.org/lwq.html
20031027 doc: qmail.7 identifies installation as netqmail and points to
         http://qmail.org/
20031027 doc: qmail-queue.8 adds explanation of $QMAILQUEUE
20031027 doc: qmail-log.5 adds reference to errors from $QMAILQUEUE script
20031027 doc: FAQ also points to http://cr.yp.to/qmail/faq.html and
         http://qmail.org/
20031027 code: qmail-smtpd identifies itself as netqmail
20031027 code: if $QMAILQUEUE is set, it's invoked instead of qmail-queue
20031024 code: changed errno from int to #include.
20031024 code: fixed .qmail parsing bug.
20031024 code: recognize 0.0.0.0 as a local address.
20031024 code: sendmail's -f flag now overrides environment variables.
19980615 version: qmail 1.03.
19980614 doc: eliminated BIN.setup in favor of a web page.
19980614 code: added other auto* to qmail-showctl output.
19980614 doc: added pointer to immhf.html in qmail-header.5.
19980614 doc: added note to TEST.receive about SMTP command format.
19980614 doc: added FAQ 5.6 on qmail-qmqpd.
19980614 code: removed unused variables in idedit.c.
19980613 code: changed conf-patrn to 002.
19980613 doc: moved SENDMAIL lower in INSTALL.
19980612 code: added install-big.
19980612 code: added BIN.Makefile.
19980612 doc: added BIN.README, BIN.setup.
19980612 code: switched to new install.
19980611 code: added idedit.
19980611 doc: added FAQ 1.3 on $QMAILMFTFILE.
19980611 doc: used bouncesaying in FAQ 5.5.
19980611 code: added except.
19980611 code: added bouncesaying.
19980611 code: allowed unbracketed IP addresses in dns_ipplus() and
         dns_mxip().
19980611 code: allowed spaces after colon in non-bracketed addresses in
         qmail-smtpd.
19980611 doc: cleaned up UPGRADE.
19980528 bug: qmail-smtpd skips first character in rcpthosts() call.
         tnx NND. impact: qmail-smtpd crashes on empty address; and it
	 allows relaying to ""@any.host. fix: use addr.s.
19980515 doc: expanded flock discussion in INSTALL.mbox.
19980515 doc: eliminated flock warning from INSTALL.maildir.
19980515 doc: split REMOVE.binmail out of INSTALL.
19980515 doc: split REMOVE.sendmail out of INSTALL.
19980515 doc: split TEST.deliver and TEST.receive out of INSTALL and
         UPGRADE.
19980515 doc: integrated INSTALL.boot into INSTALL.
19980515 code: cleaned up final output in qmail-qmqpd.c.
19980514 doc: updated procmail notes in INSTALL.mbox. tnx JRM.
19980514 doc: changed FAQ 4.4 to point to INSTALL.mbox for procmail.
         tnx JRM.
19980514 code: separated HELO and EHLO; single-line response for HELO.
         tnx to various people.
19980430 version: qmail 1.02.
19980430 doc: updated SECURITY.
19980430 doc: fixed FAQ 4.9. tnx KB.
19980430 code: changed quote2() to avoid quoting <>.
19980429 code: changed quote_need() to quote empty local parts. tnx HHO.
19980428 doc: added status notes to INSTALL and UPGRADE.
19980428 code: skip setting environment in sendmail.c if PROTO is set.
19980428 code: eliminated recipientmap.
19980428 code: added virtual users to qmail-send.c. tnx RN.
19980428 code: eliminated domain from rewrite() in qmail-send.c.
19980428 code: added binm1, binm1+df, binm2, binm2+df, binm3, binm3+df.
19980428 doc: eliminated most Mailbox references from INSTALL, UPGRADE.
19980428 code: added config-fast.
19980428 code: renamed qmail-config as config.
19980428 code: supported QMAILMFTFILE in qmail-inject.c.
19980428 code: recognized Mail-Followup-To in hfield.c.
19980428 code: replaced rwrecip() with rwappend() in qmail-inject.c.
19980428 code: cleaned up doheaderfile() in qmail-inject.c.
19980426 code: eliminated -type test from qmail-qstat to speed it up.
         tnx FT.
19980421 doc: eliminated remove-rcpthosts comments from FAQ.
19980421 doc: updated FAQ 4.3 to point to Russ Allbery's FAQ. 
19980421 doc: took account of /var/qmail/boot in INSTALL, UPGRADE, and
         INSTALL.vsm.
19980421 code: added /var/qmail/boot, with home, home+df, proc, proc+df.
19980421 doc: skipped make and make man in INSTALL.
19980420 doc: cleaned up mbox description in SENDMAIL.
19980420 code: changed QMQP port to official port 628.
19980402 doc: updated qmsmac references to fastforward.
19980402 doc: replaced qmail-upgrade man page with doc/SENDMAIL.
19980402 code: added qmqpservers output to qmail-showctl.
19980402 code: added qmail-qmqpd.
19980402 code: added qmail-qmqpc.
19980304 code: eliminated del_saywhynoexit in qmail-send.c.
19980304 code: eliminated concurrencynodel in qmail-send.c.
19980222 code: added status() to qmail-send.c.
19980222 code: added concurrencyused to qmail-send.c.
19980128 doc: added note to qmail-getpw.9 about ETXTBSY.
19980127 code: eliminated err_seenmail() in qmail-smtpd.c. tnx PO.
19980126 doc: used $DEFAULT in FAQ where possible.
19980126 code: added DEFAULT in qmail-local.
19980126 code: added -/ to qmail-pw2u.
19980126 code: revamped qmeopen() as qmesearch() with more sensible
         semantics, separating dash from ext cleanly.
19980126 code: split qmeexists() out of qmeopen() in qmail-local.c.
19980126 code: introduced safeext in qmail-local.c.

qmail-qfilter.c
/*
 * $Id: qmail-qfilter.c,v 1.25 2024-05-09 22:03:17+05:30 mbhangui Exp mbhangui $
 *
 * Copyright (C) 2001,2004-2005 Bruce Guenter <bruceg@em.ca>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <stdlib.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <unistd.h>
#include "substdio.h"
#include "env.h"
#include "scan.h"
#include "alloc.h"
#include "str.h"
#include "fmt.h"
#include "byte.h"
#include "wait.h"
#include "control.h"
#include "error.h"
#include "custom_error.h"
#include "parse_env.h"
#include "getDomainToken.h"

#ifndef TMPDIR
#define TMPDIR "/tmp"
#endif

#ifndef BUFSIZE
#define BUFSIZE 4096
#endif

#ifndef QMAIL_QUEUE
#define QMAIL_QUEUE "bin/qmail-queue"
#endif

#define QQ_OOM 51
#define QQ_PERM_MSG_REJECT 31
#define QQ_WRITE_ERROR 53
#define QQ_READ_ERROR 54
#define QQ_TEMP_MSG_REJECT 71
#define QQ_INTERNAL 81
#define QQ_BAD_ENV 91
#define QQ_DROP_MSG 99
#define QQ_WAITPID_SURPRISE  122

#define MSGIN  0
#define MSGOUT 1
#define ENVIN  3
#define ENVOUT 4
#define QQFD   5

#define memcpy(x,y,z)      byte_copy((x), (z), (y))

static const char* qqargv[2] = { QMAIL_QUEUE, 0 };

static size_t   env_len;
static size_t   msg_len;
char            strnum[FMT_ULONG];

/*
 * Parse the sender address into user and host portions
 */
size_t
parse_sender(char *env)
{
	char           *ptr = env;
	int             at, len;

	for (len = 0; *ptr != 'F' && len < env_len; len++, ptr++);
	if (*ptr != 'F')
		custom_error("qmail-qfilter", "Z", "bad envelope", 0, "X.3.0");
	ptr++;
	len++;
	if (!env_unset("QMAILNAME"))
		_exit(QQ_OOM);
	if (!*ptr) {
		if (!env_put("QMAILUSER=") || !env_put("QMAILHOST="))
			_exit(QQ_OOM);
		return 2;
	}
	at = str_rchr(ptr, '@');
	if (ptr[at]) {
		if (!env_put2("QMAILHOST", ptr + at + 1))
			_exit(QQ_OOM);
		ptr[at] = 0;
		if (!env_put2("QMAILUSER", ptr))
			_exit(QQ_OOM);
		ptr[at] = '@';
	} else {
		if (!env_put2("QMAILUSER", ptr))
			_exit(QQ_OOM);
		if (!env_put("QMAILHOST="))
			_exit(QQ_OOM);
	}
	for (; *ptr; len++, ptr++);
	return (len + 1);
}

void
parse_rcpts(char *env, int offset)
{
	size_t          len = env_len - offset;
	char           *ptr = env + offset;
	char           *buf = alloc(len);
	char           *tmp = buf;
	unsigned long   count;

	count = 0;
	while (ptr < env + env_len && *ptr == 'T') {
		size_t          rcptlen = str_len(++ptr);
		memcpy(tmp, ptr, rcptlen);
		tmp[rcptlen] = '\n';
		tmp += rcptlen + 1;
		ptr += rcptlen + 1;
		++count;
	}
	*tmp = 0;
	if (!env_put2("QMAILRCPTS", buf))
		_exit(QQ_OOM);
	strnum[fmt_ulong(strnum, count)] = 0;
	if (!env_put2("NUMRCPTS", strnum))
		_exit(QQ_OOM);
	alloc_free(buf);
}

void
parse_envelope(void)
{
	char           *env;
	size_t          offset;

	if ((env = mmap(0, env_len, PROT_READ|PROT_WRITE, MAP_PRIVATE, ENVIN, 0)) == MAP_FAILED)
		custom_error("qmail-qfilter", "Z", "mmap failed", 0, "X.3.0");
	offset = parse_sender(env);
	parse_rcpts(env, offset);
	munmap((void *) env, env_len);
}

/*
 * Create a temporary invisible file opened for read/write
 */
int
mktmpfile()
{
	char            filename[sizeof (TMPDIR) + 19] = TMPDIR "/fixheaders.XXXXXX";
	int             fd = mkstemp(filename);

	if (fd == -1)
		custom_error("qmail-qfilter", "Z", "unable to create temp file", 0, "X.3.0");
	/*
	 * The following makes the temporary file disappear immediately on
	 * program exit.
	 */
	if (unlink(filename) == -1) {
		close(fd);
		custom_error("qmail-qfilter", "Z", "unable to unlink temp file", 0, "X.3.0");
	}
	return fd;
}

/*
 * Renumber from one FD to another
 */
void
move_fd(int currfd, int newfd)
{
	if (currfd == newfd)
		return;
	if (dup2(currfd, newfd) != newfd)
		custom_error("qmail-qfilter", "Z", "move_fd: dup2 failed", 0, "X.3.0");
	if (close(currfd) != 0)
		custom_error("qmail-qfilter", "Z", "move_fd: close failed", 0, "X.3.0");
}

/*
 * Copy from one FD to a temporary FD
 */
void
copy_fd(int fdin, int fdout, size_t *var)
{
	unsigned long   bytes;
	int             tmp = mktmpfile();

	/*
	 * Copy the message into the temporary file
	 */
	for (bytes = 0;;) {
		char            buf[BUFSIZE];
		ssize_t         rd = read(fdin, buf, BUFSIZE);

		if (rd == -1)
			exit(QQ_READ_ERROR);
		if (rd == 0)
			break;
		if (write(tmp, buf, rd) != rd)
			exit(QQ_WRITE_ERROR);
		bytes += rd;
	}
	close(fdin);
	if (lseek(tmp, 0, SEEK_SET) != 0)
		custom_error("qmail-qfilter", "Z", "unable to lseek", 0, "X.3.0");
	move_fd(tmp, fdout);
	*var = bytes;
}

struct command {
	char          **argv;
	char          *env;
	struct command *next;
};
typedef struct command command;

/*
 * Split up the command line into a linked list of separate commands
 */
command        *
parse_args(int argc, char *argv[])
{
	command        *tail = NULL;
	command        *head = NULL;

	while (argc > 0) {
		command        *cmd;
		int             end = 0;
		while (end < argc && str_diff(argv[end], "--"))
			++end;
		if (end == 0)
			_exit(QQ_INTERNAL);
		argv[end] = 0;
		cmd = (command *) alloc(sizeof (command));
		if (!cmd) _exit(QQ_OOM);
		cmd->argv = argv;
		cmd->env = 0;
		cmd->next = 0;
		if (tail)
			tail->next = cmd;
		else
			head = cmd;
		tail = cmd;
		++end;
		argv += end;
		argc -= end;
	}
	return head;
}

/*
 * Read matching filter commands from control/qfilters
 */
command        *
parse_control(command* head)
{
	command        *tail;
	char           *domain;
	char           *filter;
	char           *env;
	int            ret;
	int            pos = 0;
	static stralloc filterdefs = { 0 };

	for (tail = head; tail && tail->next; tail = tail->next) ;

	if ((ret = control_readfile(&filterdefs, "control/qfilters", 0)) == -1)
		custom_error("qmail-qfilters", "Z", "unable to read qfilters", 0, "X.3.0");
	if (!ret) return head;
	if (!(domain = env_get("QMAILHOST"))) domain = "";
	while (filter = getDomainTokens(domain, &filterdefs, &pos, &env)) {
		command        *cmd;
		cmd = (command *) alloc(sizeof (command));
		if (!cmd) _exit(QQ_OOM);
		cmd->argv = (char**) alloc(4 * sizeof(char*));
		if (!cmd->argv) _exit(QQ_OOM);
		cmd->argv[0] = "/bin/sh";
		cmd->argv[1] = "-c";
		cmd->argv[2] = filter;
		cmd->argv[3] = 0;
		cmd->env = env;
		cmd->next = 0;
		if (tail)
			tail->next = cmd;
		else
			head = cmd;
		tail = cmd;
	}
	return head;
}

static void
move_unless_empty(int src, int dst, const void *reopen, size_t *var)
{
	struct stat     st;

	if (fstat(src, &st) != 0)
		_exit(QQ_INTERNAL);
	if (st.st_size > 0) {
		move_fd(src, dst);
		*var = st.st_size;
		if (reopen) {
			move_fd(mktmpfile(), src);
			if (src == ENVOUT)
				parse_envelope();
		}
	} else
	if (!reopen)
		close(src);
	if (lseek(dst, 0, SEEK_SET) != 0)
		custom_error("qmail-qfilter", "Z", "unable to lseek", 0, "X.3.0");
}

void
read_qqfd(void)
{
	struct stat     st;
	char           *buf;

	if (fstat(QQFD, &st) != 0)
		_exit(QQ_INTERNAL);
	if (st.st_size > 0) {
		if ((buf = alloc(st.st_size + 1)) == 0)
			_exit(QQ_OOM);
		if (read(QQFD, buf, st.st_size) != st.st_size)
			_exit(QQ_READ_ERROR);
		buf[st.st_size] = 0;
		qqargv[0] = buf;
	}
	close(QQFD);
}

/*
 * Run each of the filters in sequence
 */
void
run_filters(const command *first)
{
	const command  *c;
	int             i;

	move_fd(mktmpfile(), MSGOUT);
	move_fd(mktmpfile(), ENVOUT);

	for (c = first; c; c = c->next) {
		pid_t           pid;
		int             status;

		strnum[fmt_ulong(strnum, env_len)] = 0;
		if (!env_put2("ENVSIZE", strnum))
			_exit(QQ_OOM);
		strnum[fmt_ulong(strnum, msg_len)] = 0;
		if (!env_put2("MSGSIZE", strnum))
			_exit(QQ_OOM);
		if (lseek(1, 0, SEEK_SET) != 0)
			custom_error("qmail-qfilter", "Z", "unable to lseek envelope", 0, "X.3.0");
		switch ((pid = fork()))
		{
			case -1:
				_exit(121);
			case 0:
				if (c->env) parse_env(c->env);
				execvp(c->argv[0], c->argv);
				_exit(QQ_INTERNAL);
		}
		if ((i = waitpid(pid, &status, 0)) == -1) {
			strnum[fmt_ulong(strnum, errno)] = 0;
			custom_error("qmail-qfilter", "Z", "waitpid error", strnum, "X.3.0");
		}
		if (i != pid)
			_exit(QQ_WAITPID_SURPRISE);
		if (wait_signaled(status)) {
			strnum[fmt_ulong(strnum, wait_termsig(status))] = 0;
			custom_error("qmail-qfilter", "Z", "killed by signal", strnum, "X.3.0");
		}
		if (!wait_exited(status))
			custom_error("qmail-qfilter", "Z", "invalid exit status", 0, "X.3.0");
		switch (i = wait_exitcode(status))
		{
			case 0:
				break;
			case QQ_DROP_MSG:
				_exit(0);
			case QQ_PERM_MSG_REJECT:
			case 100:
				_exit(QQ_PERM_MSG_REJECT);
			case QQ_TEMP_MSG_REJECT:
			case 111:
				_exit(QQ_TEMP_MSG_REJECT);
			default:
				_exit(i);
		}
		move_unless_empty(MSGOUT, MSGIN, c->next, &msg_len);
		move_unless_empty(ENVOUT, ENVIN, c->next, &env_len);
		if (lseek(QQFD, 0, SEEK_SET) != 0)
			custom_error("qmail-qfilter", "Z", "unable to lseek qqfd", 0, "X.3.0");
	}
}

int
main(int argc, char *argv[])
{
	command  *filters;
	const char *qqq;

	filters = parse_args(argc - 1, argv + 1);

	if ((qqq = getenv("QQF_QMAILQUEUE")) != 0)
		qqargv[0] = qqq;

	copy_fd(0, 0, &msg_len);
	copy_fd(1, ENVIN, &env_len);
	parse_envelope();
	move_fd(mktmpfile(), QQFD);
   
	filters = parse_control(filters);

	run_filters(filters);
   
	read_qqfd();
	move_fd(ENVIN, 1);
	execv(qqargv[0], (char**)qqargv);
	return QQ_INTERNAL;
}

#ifndef lint
void
getversion_qmail_qfilter_c()
{
	const char     *x = "$Id: qmail-qfilter.c,v 1.25 2024-05-09 22:03:17+05:30 mbhangui Exp mbhangui $";

	x++;
}
#endif

/*
 * $Log: qmail-qfilter.c,v $
 * Revision 1.25  2024-05-09 22:03:17+05:30  mbhangui
 * fix discarded-qualifier compiler warnings
 *
 * Revision 1.24  2024-02-20 22:25:37+05:30  Cprogrammer
 * added exit code 2 (blackhole) for compatibility with qmail-multi
 *
 * Revision 1.23  2024-02-19 22:48:22+05:30  Cprogrammer
 * exit with exit code of filter
 *
 * Revision 1.22  2024-01-23 01:22:51+05:30  Cprogrammer
 * include buffer_defs.h for buffer size definitions
 *
 * Revision 1.21  2023-09-08 00:57:21+05:30  Cprogrammer
 * BUG FIX: qmail-multi, qmail-queue wasn't getting executed
 *
 * Revision 1.20  2023-03-26 08:22:44+05:30  Cprogrammer
 * fixed code for wait_handler
 *
 * Revision 1.19  2022-12-18 11:50:05+05:30  Cprogrammer
 * handle wait status with details
 *
 * Revision 1.18  2022-10-17 19:44:50+05:30  Cprogrammer
 * use exit codes defines from qmail.h
 *
 * Revision 1.17  2022-03-08 22:59:54+05:30  Cprogrammer
 * use custom_error() from custom_error.c
 *
 * Revision 1.16  2021-10-25 09:13:01+05:30  Cprogrammer
 * eliminated mktmpfd() function
 *
 * Revision 1.15  2021-06-12 18:25:31+05:30  Cprogrammer
 * removed #include "auto_qmail.h"
 *
 * Revision 1.14  2021-06-09 19:36:38+05:30  Cprogrammer
 * use qmulti() instead of exec of qmail-multi
 *
 * Revision 1.13  2021-05-29 23:49:54+05:30  Cprogrammer
 * replace str_chr with str_rchr to get domain correctly from email address
 *
 * Revision 1.12  2020-12-07 16:09:42+05:30  Cprogrammer
 * use exit code 79 for envelope format error
 *
 * Revision 1.11  2019-02-17 11:40:14+05:30  Cprogrammer
 * fixed indentation
 *
 * Revision 1.10  2016-06-03 09:58:12+05:30  Cprogrammer
 * moved qmail-multi to sbin
 *
 * Revision 1.9  2011-07-13 20:56:18+05:30  Cprogrammer
 * removed useless close()
 *
 * Revision 1.8  2011-02-17 22:11:19+05:30  Cprogrammer
 * lseek envelope fd
 *
 * Revision 1.7  2009-04-24 23:50:47+05:30  Cprogrammer
 * version 2.1 of qmail-qfilter
 *
 * Revision 1.6  2009-04-22 20:04:30+05:30  Cprogrammer
 * added customized error messages
 *
 * Revision 1.5  2009-04-21 20:43:38+05:30  Cprogrammer
 * use bin/qmail-multi instead of absolute path
 *
 * Revision 1.4  2009-04-02 20:35:42+05:30  Cprogrammer
 * call qmail-multi instead of qmail-queue
 *
 * Revision 1.3  2004-10-22 20:28:40+05:30  Cprogrammer
 * added RCS id
 *
 * Revision 1.2  2004-06-20 17:33:08+05:30  Cprogrammer
 * use setenv() for mysetenv()
 * use mkstemp() instead of tempnam()
 *
 * Revision 1.1  2004-05-23 20:28:44+05:30  Cprogrammer
 * Initial revision
 *
 */
